### ISSUE TITLE
Rationalize namespaces

### ISSUE THREAD

**Issue by rdeago** [unosquare/embedio#299]
I propose to have namespaces related to API functionality, rather than code items.

My rationale is that namespaces should, in general, reflect sets of functionality: routing, serving files, WebSocket handling, web APIs, and so on.

The `EmbedIO.Modules` namespace is currently polluted with base types and support types pertaining to only one kind of module: `BasicAuthenticationModuleBase`, `BasicAuthenticationModuleExtensions`, `FileModuleBase`, `FileModuleBaseExtensions`, `WebApiController`...

I already created the `EmbedIO.Routing` namespace because I saw I was spreading routing-related functionality over three or four different namespaces, all to be `using`-ed by anyone using a routing or Web API module.

Some possible namespaces include:
* `EmbedIO.Sessions`
* `EmbedIO.Files`
* `EmbedIO.Authentication`
* `EmbedIO.Actions` (for `ActionModule` and `RedirectModule`)
* `EmbedIO.Cors`
* `EmbedIO.WebApi`
* `EmbedIO.WebSocket` (to mirror the WebSocket protocol's name) or `EmbedIO.WebSockets`(to make life easier for people accustomed to `System.Net.WebSockets`)

`WebModuleBase` and `ModuleGroup`, being part of the basic infrastructure of a web server, should be moved to the `EmbedIO` namespace.

I also propose that:

* the `EmbedIO.Constants` namespace be suppressed and its types moved to relevant namespaces;

* EmbedIO's HTTP listener implementation be moved to a `EmbedIO.Net.HttpListener` namespace;

* types pertaining to networking, such as `HttpHeaderNames` and `HttpStatusDescription`, be moved to `EmbedIO.Net`.

---

To summarize the rationale into a quick guide to namespaces for contibutors (to be probably included in the contributing guide if this issue gets enough love):

* Namespaces shall, as a general guideline, group types by implemented functionality.

* To make source browsing easier, `internal` types used by types in namespace `EmbedIo.Foo` shall be in their own `EmbedIO.Foo.Internal` namespace.

* The `EmbedIO.Utilities` namespace contains general-purpose utility types that could, in theory, be used outside the context of EmbedIO. Types in this namespace that prove generally useful may be moved or copied to SWAN and removed from EmbedIO in a future major version.

---

Feedback is, as always, highly appreciated.

**Comment by geoperez** [comment]
I'm fine changing the namespaces, we are introducing a lot of breaking changes in the next version. So namespace is a trivial side effect.

**Comment by rdeago** [comment]
Great! Will do.

**Comment by rdeago** [comment]
Closed by 32f909f542a2f4d77a138c141bd9d93426f70370.

Changes are summarized below:

* Refactor namespaces as per original post.
* Prepend `System` to the names of `System.Net` façades.
* Rename `IHttpBase` to a more specific `IHttpMessage`.
* Move WebSocket-related types to `EmbedIO.WebSockets`:
  * the final "s" is for consistency with `System.Net.WebSockets`;
  * not `EmbedIO.Net.WebSockets` because WebSocket support being a feature of EmbedIO takes precedence over it being a network protocol.
* Move static classes providing network-related constants (`HttpHeaderNames` et al.) to the `EmbedIO` namespace:
  * not `EmbedIO.Net` because it would have been one more namespace to use almost everywhere `EmbedIO` is used.
* Make `CompressionMethods` its own class instead of a nested class of `HttpHeaderNames`. It made sense when `HttpHeaderNames` was called `HttpHeaders`, but now it stands better on its own.
* Make some internal-use classes `internal` instead of `public`.
* Move `TestHttpRequest` and `TestHttpResponse` to `EmbedIO.Tests` since they are not `internal` any longer.
* Clean up `using` directives in all `EmbedIO` project.

**Comment by geoperez** [comment]
Thanks for the summary!
